_posts/2019-05-31-ucx-dgx.md

```diff
@@ -0,0 +1,531 @@
+                "plugins": [CPUAffinity(affinity[i])],
+                "silence_logs": silence_logs,
+                "memory_limit": memory_limit,
+            },
+        }
+        for ii, i in enumerate(CUDA_VISIBLE_DEVICES)
+    }
+
+    scheduler = {
+        "cls": Scheduler,
+        "options": {
+            "interface": interface + str(CUDA_VISIBLE_DEVICES[0] // 2),
+            "protocol": "ucx",
+            "dashboard_address": dashboard_address,
+        },
+    }
+
+    return SpecCluster(
+        workers=spec, scheduler=scheduler, silence_logs=silence_logs, **kwargs
+    )
+```
+
+However, we never got the NVLink structure down.  The Dask scheduler currently
+still assumes uniform bandwidths between workers.  We've started to make small
+steps towards changing this, but we're not there yet (this will be useful as
+well for people that want to think about in-rack or cross-data-center
+deployments).
+
+As usual, in solving a highly specific problem, we were able to solve a number
+of lingering general features, which then made our specific problem easy to
+write down.
+
+
+Future Work
+-----------
+
+-   **Routing within complex networks**:
+    If you restrict yourself to four of the eight GPUs in a DGX, you can get 5-10 GB/s
+    between pairs of GPUs.  For some workloads this can be significant.  It
+    makes the system feel much more like a single unit than a bunch of isolated
+    machines.
+
+    However we still can't get great performance across the whole DGX because
+    there are many GPU-pairs that are not connected by NVLink, and so we get 10x
+    slower speeds.  These dominate communication costs if you naively try to use
+    the full DGX.
+
+    This might be solved either by:
+
+    1.  Teaching Dask to avoid these communications
+    2.  Teaching UCX to route communications like these through a chain of
+        multiple NVLink connections
+    3.  Avoiding complex networks altogether.  Apparently newer systems like
+        the DGX-2 use NVSwitch, which is more uniform.
+
+-   **CPU:** We can get 1-2 GB/s across infiniband, which isn't bad, but also
+    wasn't the full 5-8 GB/s that we was hoping for.  This deserves more serious
+    profiling to determine what is going wrong.  The current guess is that this
+    has to do with memory allocations.
+
+    ```python
+    In [1]: %time _ = b'0' * 1000000000  # 1 GB
+    CPU times: user 248 ms, sys: 223 ms, total: 472 ms
+    Wall time: 470 ms   # <<----- Around 2 GB/s.  Slower than I expected
+    ```
+
+    Probably we're just doing something dumb here.
+
+-   **Package UCX:** Currently I'm building the UCX and UCX-Py libraries from
+    source (see appendix below for instructions).  Ideally these would become
+    conda packages.  [John Kirkham](https://github.com/jakirkham) (Conda Forge,
+    NVIDIA, Dask) is taking a look at this along with the UCX developers from
+    Mellanox.
+
+-   **Learn Heterogeneous Bandwidths:** In order to make good scheduling
```

Comment:
Do we have a distributed issue about this?

Reply:
https://github.com/dask/distributed/issues/2743